test/Common/Test.hs

```diff
@@ -874,13 +874,57 @@ testSelectSubQuery run = do
                 on $ lord ^. LordId ==. deed ^. DeedOwnerId
                 return (lord ^. LordId, deed ^. DeedId)
 
-        (ret :: [(Value (Key Lord), Value Int)]) <- select $ fromQuery q $ \(lordId,deedId) -> do
+            q' = from $ \lords -> do
+                 (lordId, deedId) <- innerJoinQuery q
+                 on (lordId ==. lords ^. LordId)
                  groupBy (lordId)
                  return (lordId, count deedId)
 
+        (ret :: [(Value (Key Lord), Value Int)]) <- select q'
         liftIO $ ret `shouldMatchList` [ (Value l3k, Value 7)
                                        , (Value l1k, Value 3) ]
 
+    it "supports LEFT OUTER JOIN" $ do
+      run $ do
+        p1e <- insert' p1
+        p2e <- insert' p2
+        p3e <- insert' p3
+        p4e <- insert' p4
+        b12e <- insert' $ BlogPost "b" (entityKey p1e)
+        b11e <- insert' $ BlogPost "a" (entityKey p1e)
+        b31e <- insert' $ BlogPost "c" (entityKey p3e)
+        let q = from $ \p -> do
+                 mb <- leftOuterJoinQuery $ from $ \(mb_ :: SqlExpr (Entity BlogPost)) -> pure mb_ -- dummy subquery that just returns all the values in BlogPost table
+                 on (just (p ^. PersonId) ==. mb ?. BlogPostAuthorId)
+                 orderBy [ asc (p ^. PersonName), asc (mb ?. BlogPostTitle) ]
+                 return (p, mb)
+        ret <- select q
+        liftIO $ ret `shouldBe` [ (p1e, Just b11e)
+                                , (p1e, Just b12e)
+                                , (p4e, Nothing)
+                                , (p3e, Just b31e)
+                                , (p2e, Nothing) ]
+
+    it "supports joining two queries" $ do
+      run $ do
+        p1e <- insert' p1
+        p2e <- insert' p2
+        b12e <- insert' $ BlogPost "b" (entityKey p1e)
+        b11e <- insert' $ BlogPost "a" (entityKey p1e)
+        let q = from $ \p -> do
+                 mb <- leftOuterJoinQuery $ from $ \(mb_ :: SqlExpr (Entity BlogPost)) -> pure mb_ -- dummy subquery that just returns all the values in BlogPost table
+                 mb' <- leftOuterJoinQuery $ from $ \(mb_ :: SqlExpr (Entity BlogPost)) -> pure mb_ -- dummy subquery that just returns all the values in BlogPost table
+                 on (just (p ^. PersonId) ==. mb ?. BlogPostAuthorId)
+                 on (just (p ^. PersonId) ==. mb' ?. BlogPostAuthorId)
+                 orderBy [ asc (p ^. PersonName), asc (mb ?. BlogPostTitle) ]
+                 return (p, mb, mb')
```

Comment:
Ok, so if I understand right, this translates to this SQL query:

```sql
SELECT p.*, mb1.*, mb2.*
FROM person AS p
LEFT JOIN (SELECT * FROM blog_post) AS mb1
LEFT JOIN (SELECT * FROM blog_post) AS mb2
ORDER BY p.name ASC, mb1.title ASC
```

Which isn't taking advantage of the power of the code. We could add additional filters, or return `Value x` instead of `Entity x`. Intriguing.

I think we can avoid the partial stuff by having the basic API be something like:

```haskell
joinSubquery
  :: SqlExpr lhs
  -- ^ The table we're joining to. Or maybe it should be a value?
  -> JoinKind
  -> SqlQuery (SqlExpr rhs)
  -- ^ The query that returns the thing we care about
  -> (SqlExpr rhs -> SqlExpr Bool)
  -- ^ The on clause.
  -> SqlQuery lhs
```

So then we'd rewrite this code as:

```haskell
select $
from $ \p -> do
mb <- joinSubquery 
  p 
    LeftOuterJoin
      (from pure :: SqlQuery (Entity Blogpost))
      (\mb -> p ^. PersonId ==. mb ?. BlogPostAuthorId)
mb' <- -- etc.
...
```

With this API, we have the `on` clause immediately. This is appealing to me because we don't have to do any fancy business to align `on` clauses, and I think we can avoid the partiality and extra `Monoid` business entirely.

I think that I would prefer to eventually discard the `from` magic entirely, writing something like:

```haskell
select $ do
  p <- from Person
  mb <- join p LeftOuter (\bp -> bp ^. BlogPostAuthorId ==. p ^. PersonId)
  pure (p ^. PersonName, bp ?. BlogPostTitle)
```

This would give us a type like:

```haskell
join
  :: SqlExpr lhs
  -> JoinKind
  -> (SqlExpr rhs -> SqlExpr bool)
  -> SqlQuery (SqlExpr rhs') -- where rhs' is somehow determined by the value of `JoinKind`... oh no singletons
```

Reply:
I am not sure that we can switch to your form of the api without also changing how joins work completely. It is not clear to me how your api supports multiple joins in the left hand side of the join tree. 

As far as simplifying to a tuple of the join tree and the on clauses that could help potentially. The partial type was implemented because I couldn't find a good way to support the lhs. Its the same reason why this implementation doesnt support nested joins against subqueries, though if nested queries are little used I can't imagine anyone wanting for them. 

As far as the constraint on the return value its pretty simple, they need to be aliasable types using ToAlias and ToAliasReference, the same as with fromSubquery.